### PR TITLE
Migrate away from ember-cli-head for <title> updates

### LINKED ISSUE #167: remove dependency on ember-cli-head
TODO: Change `ember-page-title` to remove the `ember-cli-head` requirement (for both SSR and in browser)
Teams will have to add the ember-cli-head addon if they want the functionality from both.

This will unblock the [related RFC](https://github.com/emberjs/rfcs/pull/645)

We'll need to create a new release and do a major version bump. 

Notes: 
After this is finished, update RFC 645 to add both ember-page-title and ember-cli-head as dev dependencies of apps.

### PR DESCRIPTION
**Pull request by raido**
Use direct page title manipulation instead of ember-cli-head.

Fastboot apps have to remove `<title>` from app/index.html as before and for non-fastboot apps everything should work out of the box.

Closes #167

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -3,7 +3,7 @@ language: node_js
 node_js:
   # we recommend testing addons with the same minimum supported node version as Ember CLI
   # so that your addon works for all apps
-  - "8"
+  - "10"
```
Comment (rwjblue) [inline .travis.yml:6]: Let's do this in a separate PR specifically focused on dropping node < 10.

File: README.md [hunk 2]
```diff
@@ -54,21 +52,15 @@ module.exports = function (environment) {
...
-This file is added automatically if you use `ember install`. This is for all the folks using ember-cli-head addons like ember-cli-meta-tags.
-
 ### Deprecations
 
 - Since **v5.2.2**: The `{{title}}` helper has been deprecated, use `{{page-title}}` instead, it has the same API. The
   `{{title}}` helper was an AST transform and will be removed in the next major release.
 
+### Upgrading notes for 5.x to 6.x
+
+From 6.x onward, you don't need to have `{{head-layout}}` within your application's `application.hbs` template. If you have `ember-cli-head` as direct dependency on your project, but not used for anything else you can also remove it.
```
Comment (rwjblue) [inline README.md:62]: Maybe mention something like:

> ember-page-title no longer requires the usage of ember-cli-head

At the beginning of the existing prose?

File: addon/helpers/page-title.js [hunk 3]
```diff
@@ -1,13 +1,30 @@
...
+  const toBeTitle = tokens.toString();
+
+  if (isFastBoot) {
+    let childNodes = document.head.childNodes;
+    for (let i = 0; i < childNodes.length; i++) {
+      let node = childNodes.item(i);
+      if (node.tagName.toLowerCase() === 'title') {
+        document.head.removeChild(node);
+      }
+    }
```
Comment (rwjblue) [inline addon/helpers/page-title.js:20]: Why do we need to remove the prior `<title>` tags that we find? In FastBoot the only thing that would be here that might include the prior `<title>` would have been someones `head.hbs` or something, the static `<title>` that is included in `app/index.html` would **not** be overridden by this.

File: addon/helpers/page-title.js [hunk 4]
```diff
@@ -18,7 +35,9 @@ function updateTitle(tokens) {
  */
 export default Helper.extend({
   pageTitleList: service(),
-  headData: service(),
+  document: service('-document'),
+  fastboot: service(),
+  isFastBoot: reads('fastboot.isFastBoot'),
```
Comment (rwjblue) [inline addon/helpers/page-title.js:40]: I think we can't rely on `service:fastboot` being present. Applications that don't have `FastBoot` installed at all will throw an error here.

File: addon/helpers/page-title.js [hunk 5]
```diff
@@ -40,7 +59,7 @@ export default Helper.extend({
     hash.id = guidFor(this);
     hash.title = params.join('');
     tokens.push(hash);
-    scheduleOnce('afterRender', get(this, 'headData'), updateTitle, tokens);
+    scheduleOnce('afterRender', updateTitle, get(this, 'document'), tokens, get(this, 'isFastBoot'));
```
Comment (rwjblue) [inline addon/helpers/page-title.js:62]: Instead of passing in `isFastBoot` to the function, lets have the function directly do:

```js
let isFastBoot = typeof FastBoot !== 'undefined';
```

File: addon/helpers/page-title.js [hunk 6]
```diff
@@ -52,16 +71,17 @@ export default Helper.extend({
     let router = getOwner(this).lookup('router:main');
     let routes = router._routerMicrolib || router.router;
     let { activeTransition } = routes || {};
-    let headData = get(this, 'headData');
+    let document = get(this, 'document');
+    let isFastBoot = get(this, 'isFastBoot');
     if (activeTransition) {
       activeTransition.promise.finally(function () {
-        if (headData.isDestroyed) {
+        if (document.isDestroyed) {
```
Comment (rwjblue) [inline addon/helpers/page-title.js:78]: I don't think this will ever be set? We probably have to do `this.isDestroying` here

File: tests/acceptance/posts-test.js [hunk 7]
```diff
@@ -1,95 +1,89 @@
...
-    let element = document.querySelector('head title');
-    return element && element.innerText.trim().replace(/^\(\d+\/\d+\)/, '');
-  }
-
   test('the default configuration works', async function (assert) {
     assert.expect(1);
     await visit('/posts');
 
-    assert.equal(title(), 'Posts | My App');
+    assert.equal(getPageTitle(window.document), 'Posts | My App');
```
Comment (rwjblue) [inline tests/acceptance/posts-test.js:13]: Can we default this param (instead of passing it all the time)?

File: tests/dummy/app/index.html [hunk 8]
```diff
@@ -1,7 +1,7 @@
 <!DOCTYPE html>
 <html>
   <head>
-    <title>Initial page title - removed by addon initializer</title>
+    <!-- <title>Initial page title - removed by addon initializer</title> -->
```
[the remark this answers is not in the document]
Comment (rwjblue) [reply tests/dummy/app/index.html:4]: I think apps will always have to remove this, we should remove it for them in their blueprint. And in the RFC that suggests adding ember-page-title to all new apps we should explicitly state that this will be removed in favor of actually invoking `{{title}}` in the `application.hbs`.

File: tests/helpers/get-page-title.js [hunk 9]
```diff
@@ -0,0 +1,7 @@
+// Testem appends progress to the title...
```
Comment (rwjblue) [inline tests/helpers/get-page-title.js:1]: Ugh, that sucks 😭

File: tests/fastboot/title-test.js [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+  });
+
+  test('multitoken titles work', async function (assert) {
+    assert.expect(1);
+    let { htmlDocument } = await visit('/feeds/tomster');
+
+    assert.equal(getPageTitle(htmlDocument), 'Tomster (@tomster)');
+  });
+
+  /*test('loading substates are not shown', async function (assert) {
```
[the remark this answers is not in the document]
Comment (rwjblue) [reply tests/fastboot/title-test.js:65]: In _general_ I would not expect to actually have loading substates sent down the wire in a FastBoot context (FastBoot on the server would just wait the full time).

File: tests/dummy/app/index.html (hunk 8 above)
[the remark this answers is not in the document]
Comment (rwjblue) [reply tests/dummy/app/index.html:4]: FWIW, I'm pretty sure that this was _already_ a constraint and not something new. Users of existing ember-page-title releases that are using ember-cli-fastboot along with ember-cli-head would have had to remove `<title>` from `app/index.html`.

File: addon/helpers/page-title.js (hunk 3 above)
Comment (raido) [reply addon/helpers/page-title.js:20]: It is not about the index.html static `title` tag, its about nested routes page-title invocations. We either lookup existing `title` element or clean up them all, so nested route titles like `A | B | C` - would render correctly.

What we can do is lookup existing `title` element we add on our own and remove all textNodes from it but end result is kind of the same.

File: addon/helpers/page-title.js (hunk 4 above)
Comment (raido) [reply addon/helpers/page-title.js:40]: Yes, that is correct. Why I ended up with this is for draft was that I wanted `ember test` to run both fastboot and browser tests in one go. Once I refactor this, those tests need to be run separately.

File: tests/acceptance/posts-test.js (hunk 7 above)
Comment (raido) [reply tests/acceptance/posts-test.js:13]: We can.

File: .travis.yml (hunk 1 above)
Comment (raido) [reply .travis.yml:6]: Definitely, together with general dependencies bump.

File: tests/dummy/app/index.html (hunk 8 above)
[the remark this answers is not in the document]
Comment (raido) [reply tests/dummy/app/index.html:4]: Wait, how we can this? If app/index.html has no `<title>` then search engines index pages without titles out of the box. All non fastboot apps that today set meaningful static title for search engines will have conflicting setup out of box.

File: addon/helpers/page-title.js (hunk 3 above)
Comment (rwjblue) [reply addon/helpers/page-title.js:20]: Hmm, maybe we can track the current element and just update its contents on subsequent invocations?

File: addon/helpers/page-title.js (hunk 4 above)
Comment (rwjblue) [reply addon/helpers/page-title.js:40]: Maybe make a `ember-try.js` scenario for `fastboot-tests` that adds the required dependencies? That way when `ember try:each` runs it will execute them (then you'd guard the `ember-cli-fastboot-testing` related tests for when `typeof FastBoot` isn't present)...

File: addon/helpers/page-title.js (hunk 3 above)
Comment (raido) [reply addon/helpers/page-title.js:20]: https://github.com/ember-fastboot/simple-dom/blob/master/packages/%40simple-dom/document/src/node.ts#L38 - I have not found any update methods on simple-dom `node` implementation to do this.

Comment (raido) [reply addon/helpers/page-title.js:20]: I don't think we can track either, because if we imagine nested routes:

`A -> B -> C ... n`

Then each of them has `page-title` invocation in template, each of them triggers `run.schedule afterRender` queueing and the fact that the update method is separate method here, is only thing that helps runloop to run it just once, although N number of invocations tried to initiate it again before afterRender queue ran.

Not sure how we could keep track of the element if don't know which of the helper instances has latest element reference.

File: addon/helpers/page-title.js (hunk 4 above)
Comment (raido) [reply addon/helpers/page-title.js:40]: Yes, we'll end up with that. 👍

File: addon/helpers/page-title.js (hunk 6 above)
Comment (raido) [reply addon/helpers/page-title.js:78]: Actually be the time this callback runs 'this' is already undefined. With tests we have at the moment. That's why reference are stored before finally callback definition. It's the destroy() method here not willDestroy().

File: addon/helpers/page-title.js (hunk 3 above)
Comment (knownasilya) [reply addon/helpers/page-title.js:20]: maybe we need to put the update logic into a service and the helper only updates some datastructure that keeps the order and builds out a string

File: .travis.yml (hunk 1 above)
Comment (raido) [reply .travis.yml:6]: PR for upgrade is here - https://github.com/adopted-ember-addons/ember-page-title/pull/170

File: tests/dummy/app/index.html (hunk 8 above)
[the remark this answers is not in the document]
Comment (rwjblue) [reply tests/dummy/app/index.html:4]: I suppose we could have ember-cli-fastboot's blueprint remove it? But for the most part, isn't it safe to assume the search engine will run our JS nowadays?

[the remark this answers is not in the document]
Comment (raido) [reply tests/dummy/app/index.html:4]: Not sure about. 

https://html.spec.whatwg.org/multipage/semantics.html#the-title-element - it does not say directly title is mandatory but seems strongly recommended.

https://validator.w3.org/ - from this tool perspective page without title is not valid HTML page.

File: addon/helpers/page-title.js (hunk 6 above)
Comment (rwjblue) [reply addon/helpers/page-title.js:78]: This will never be ￼defined here, need to change to arrow function.

Comment (raido) [reply addon/helpers/page-title.js:78]: Oh my, I completely missed that. Yeah, then it will work just fine.

**Comment by rwjblue** [comment]
Ready for a rebase now

**Comment by raido** [comment]
It is rebased and I will continue addressing PR feedback after CI run.

**Comment by raido** [comment]
Interesting, it seems by adding ember try scenario and moving fastboot related dependencies out of package.json - previously failing `beta` and `canary` scenarios began to pass again - https://travis-ci.org/github/adopted-ember-addons/ember-page-title/builds/721020282 🤔

File: addon/helpers/page-title.js (hunk 4 above)
Comment (raido) [reply addon/helpers/page-title.js]: Scenario added.

File: addon/helpers/page-title.js [hunk 11]
```diff
@@ -1,13 +1,30 @@
...
-  set(this, 'title', tokens.toString());
+function updateTitle(document, tokens, isFastBoot) {
+  const toBeTitle = tokens.toString();
+
+  if (isFastBoot) {
+    let childNodes = document.head.childNodes;
+    for (let i = 0; i < childNodes.length; i++) {
+      let node = childNodes.item(i);
+      if (node.tagName.toLowerCase() === 'title') {
+        document.head.removeChild(node);
```
[the remark this answers is not in the document]
Comment (raido) [reply addon/helpers/page-title.js:18]: Seems to be irrelevant after general dependencies upgrade.

**Comment by raido** [comment]
I removed ember-try fastboot scenario because the "typeof FastBoot" check allows us to run all other scenarios with browser + fastboot for more coverage.

File: addon/services/page-title-list.js [hunk 12]
```diff
@@ -190,6 +195,26 @@ export default Service.extend({
     }
   }),
 
+  // TODO use something else for initial render for routes
+  // without model() returning promise
+  // Example: A -> B -> C - when C is the target route and model() takes long time to load
+  // title of A and B will no be rendered until activeTransition.promise resolves.
```
Comment (raido) [inline addon/services/page-title-list.js:201]: This is actually not true and I'll clean this up.  For browser tests there's no difference if we wait for promise.finally or not. For fastboot there is. But since with fastboot for nested routes user anyway sees final result once page is fully returned from server, then it is not possible to see intermediate state. 

And due to waiting for the promise, our update logic for fastboot is now simple title tag adding to head without removing existing tags etc. The previous mess of for loop was required because we eagerly rendered tokens even in mid transitions.

**Comment by raido** [comment]
Beta and Canary failure seems to be coming from here:

https://github.com/glimmerjs/glimmer-vm/blob/6b0db3aaa9688c2906e0f2b06baed187e9991129/packages/%40glimmer/runtime/lib/dom/sanitized-values.ts#L43

Not sure how and why this happens though. Any ideas @rwjblue?

```
/var/folders/jp/2wh258pj23xg7sqgdf8g7gnc0000gn/T/broccoli-82322jRSxGVWj5u1V/out-183-append_ember_auto_import_analyzer/assets/vendor.js:54840
    var nodeURL = _nodeModule.module.require('url');
                                     ^

TypeError: Cannot read property 'require' of null
    at /var/folders/jp/2wh258pj23xg7sqgdf8g7gnc0000gn/T/broccoli-82322jRSxGVWj5u1V/out-183-append_ember_auto_import_analyzer/assets/@glimmer/runtime.js:288:1
    at internalRequire (/var/folders/jp/2wh258pj23xg7sqgdf8g7gnc0000gn/T/broccoli-82322jRSxGVWj5u1V/out-183-append_ember_auto_import_analyzer/assets/loader.js:54:1)
    at internalRequire (/var/folders/jp/2wh258pj23xg7sqgdf8g7gnc0000gn/T/broccoli-82322jRSxGVWj5u1V/out-183-append_ember_auto_import_analyzer/assets/loader.js:50:1)
    at internalRequire (/var/folders/jp/2wh258pj23xg7sqgdf8g7gnc0000gn/T/broccoli-82322jRSxGVWj5u1V/out-183-append_ember_auto_import_analyzer/assets/loader.js:50:1)
    at internalRequire (/var/folders/jp/2wh258pj23xg7sqgdf8g7gnc0000gn/T/broccoli-82322jRSxGVWj5u1V/out-183-append_ember_auto_import_analyzer/assets/loader.js:50:1)
    at require (/var/folders/jp/2wh258pj23xg7sqgdf8g7gnc0000gn/T/broccoli-82322jRSxGVWj5u1V/out-183-append_ember_auto_import_analyzer/assets/loader.js:87:1)
    at /var/folders/jp/2wh258pj23xg7sqgdf8g7gnc0000gn/T/broccoli-82322jRSxGVWj5u1V/out-183-append_ember_auto_import_analyzer/assets/vendor.js:72924:1
    at /var/folders/jp/2wh258pj23xg7sqgdf8g7gnc0000gn/T/broccoli-82322jRSxGVWj5u1V/out-183-append_ember_auto_import_analyzer/assets/vendor.js:72925:2
    at Script.runInContext (vm.js:131:20)
```

File: tests/dummy/app/index.html (hunk 8 above)
[the remark this answers is not in the document]
Comment (raido) [reply tests/dummy/app/index.html]: I made tests assertions smarter to cope with 2 title tags being present for Fastboot tests with knowledge that in real life we'll ask people or make ember-cli-fastboot blueprint remove title from index.html.

File: addon/helpers/page-title.js (hunk 3 above)
Comment (raido) [reply addon/helpers/page-title.js]: This actually will not be a problem if we properly wait for transition to finish for all title updates.

File: README.md (hunk 2 above)
Comment (raido) [reply README.md]: I simplified it.

**Comment by raido** [comment]
The test / dummy app has few bugs still from previous times already. It's broken on Github pages current version as well, I'll see if it is an easy fix or evolves separate PR after this one.

**Comment by raido** [comment]
Canary is green now, beta bump is not available yet, therefore it's still failing.

**Comment by rwjblue** [comment]
Restarted the beta CI job

**Comment by raido** [comment]
@rwjblue any updates on review here?

**Comment by raido** [comment]
Any updates here?

File: addon/services/page-title-list.js [hunk 13]
```diff
@@ -190,6 +192,22 @@ export default Service.extend({
     }
   }),
 
+  scheduleTitleUpdate() {
+    let router = getOwner(this).lookup('router:main');
+    let routes = router._routerMicrolib || router.router;
```
Comment (rwjblue) [inline addon/services/page-title-list.js:197]: Mind adding a small inline comment here to indicate when we can remove the `|| router.router` fallback? It'll make cleanup easier when we drop support for that Ember version..

Comment (raido) [reply addon/services/page-title-list.js:197]: Hmm. Travis is configured for 3.12+. Technically it could be dropped now?

We are going to so major version bump anyway, so maybe worth to drop it in one go?

Comment (rwjblue) [reply addon/services/page-title-list.js:197]: Ya, lets do that.

Comment (raido) [reply addon/services/page-title-list.js]: Cleaned up.

File: addon/services/page-title-list.js [hunk 14]
```diff
@@ -205,12 +222,26 @@ export default Service.extend({
...
 
+  _updateTitle() {
+    const toBeTitle = this.toString();
+
+    if (isFastBoot) {
+      // in fastboot context "document" is instance of ember-fastboot/simple-dom document
+      let titleEl = this.document.createElement('title');
+      let titleContents = this.document.createTextNode(toBeTitle);
+      titleEl.appendChild(titleContents);
+      this.document.head.appendChild(titleEl);
```
Comment (knownasilya) [inline addon/services/page-title-list.js:233]: Could this end up in a scenario where there is more than one title? Esp if there was one already?

**Review by knownasilya — CHANGES_REQUESTED** [review]
A few questions, or potential changes

File: tests/helpers/get-page-title.js [hunk 15]
```diff
@@ -0,0 +1,10 @@
...
+// and there's no way to stop this at the moment
+
+export function getPageTitle(doc) {
+  // In Fastboot context we get 2 title elements if we don't remove one from app/index.html
+  // In real world applications, it is mandatory to remove <title> from app/index.html
+  // We are keeping both for sake for testing browser and fastboot scenarios
+  let element = [...(doc || window.document).querySelectorAll('head title')].pop();
+  return element && element.innerText.trim().replace(/^\(\d+\/\d+\)/, '');
```
[the remark this answers is not in the document]
Comment (raido) [reply tests/helpers/get-page-title.js:9]: This, this was here before me, not my addition - https://github.com/adopted-ember-addons/ember-page-title/pull/168#discussion_r475844811

File: addon/services/page-title-list.js (hunk 14 above)
Comment (raido) [reply addon/services/page-title-list.js:233]: `ember-cli-fastboot` blueprint will/should be updated - so that upon install default index.html <title> is replaced with title in application.hbs. Bascially for Fastboot requirements about title element removal are the same that are already in README of this repository.

Even if there are more than one title element, in Fastboot context SimpleDOM document has now access to mutate original HTML.

Comment (raido) [reply addon/services/page-title-list.js:233]: We could add linter rule for "no-fastboot-index-title" eventually as well. 🤔

**Review by knownasilya — APPROVED** [review]

File: addon/services/page-title-list.js (hunk 14 above)
Comment (rwjblue) [reply addon/services/page-title-list.js:233]: > ember-cli-fastboot blueprint will/should be updated - so that upon install default index.html <title> is replaced as title in application.hbs. Basically for Fastboot requirements about title element removal are the same that are already in README of this repository.

I think both blueprints (ember-cli-fastboot and ember-page-title) could do the check to see if both `ember-cli-fastboot` and `ember-page-title` are present and if so remove the `<title></title>` from `app/index.html`. 

Mind making a meta issue for that here, and cross linking in an issue in `ember-cli-fastboot`?

File: tests/helpers/get-page-title.js (hunk 15 above)
[the remark this answers is not in the document]
Comment (rwjblue) [reply tests/helpers/get-page-title.js:9]: @knownasilya - its basically due to testem mucking with the title during test runs

**Review by rwjblue — APPROVED** [review]

**Comment by rwjblue** [comment]
Thank you so much for pushing this to completion @raido!

File: addon/services/page-title-list.js (hunk 14 above)
Comment (raido) [reply addon/services/page-title-list.js:233]: Issues created: 

* https://github.com/adopted-ember-addons/ember-page-title/issues/174
* https://github.com/ember-fastboot/ember-cli-fastboot/issues/792
